p2p/smc_service.py

```diff
@@ -0,0 +1,153 @@
+            raise ValueError("Web3 connection has not been configured yet")
+
+        super().__init__(token)
+        self.smc_address = smc_address
+        self.private_key = private_key
+
+        self._smc_handler = self.w3.eth.contract(
+            address=self.smc_address,
+            ContractFactoryClass=SMCHandler,
+            config=self.config,
+        )
+        self._log_handler = LogHandler(self.w3, self.config["PERIOD_LENGTH"])
+
+        self._shard_trackers: Dict[int, ShardTracker] = {}
+        self._added_header_queues: DefaultDict[
+            int,
+            List[asyncio.Queue[CollationHeader]]
+        ] = collections.defaultdict(list)
+        for shard_id in shard_ids or []:
+            self.start_watching_shard(shard_id)
+
+        self._processed_header_hashes: Set[Hash32] = set()
+
+    @property
+    def add_header_gas_price(self) -> int:
+        return 11 * 1000 * 1000 * 1000
+
+    def start_watching_shard(self, shard_id: int) -> asyncio.Queue[CollationHeader]:
+        if shard_id in self._shard_trackers:
+            raise ValueError("Already watching shard {}".format(shard_id))
+
+        self.logger.debug("Start watching shard {}".format(shard_id))
+        shard_tracker = ShardTracker(self.config, shard_id, self._log_handler, self.smc_address)
+        queue: asyncio.Queue[CollationHeader] = asyncio.Queue()
+        self._shard_trackers[shard_id] = shard_tracker
+        self._added_header_queues[shard_id].append(queue)
+        return queue
+
+    def stop_watching_shard(self, shard_id: int) -> None:
+        if shard_id not in self._shard_trackers:
+            raise ValueError("Not watching shard {}".format(shard_id))
+
+        self.logger.debug("Stop watching shard {}".format(shard_id))
+        self._shard_trackers.pop(shard_id)
+        self._added_header_queues.pop(shard_id)
+
+    def watching_shard(self, shard_id: int) -> bool:
+        return shard_id in self._shard_trackers
+
+    async def _run(self) -> None:
+        while True:
+            for shard_id, shard_tracker in self._shard_trackers.items():
+                added_header_logs = await shard_tracker.get_add_header_logs()
+                for log in added_header_logs:
+                    proposer = self._smc_handler.get_collation_proposer(log.period, log.shard_id)
+                    header = CollationHeader(
+                        shard_id=log.shard_id,
+                        period=log.period,
+                        chunk_root=log.chunk_root,
+                        propser_address=proposer,
+                    )
+
+                    if header.hash in self._processed_header_hashes:
+                        continue
+
+                    self.logger.debug("Got AddHeader log for header %s", header)
+                    self._processed_header_hashes.add(header.hash)
+                    for added_header_queue in self._added_header_queues[shard_id]:
+                        await added_header_queue.put(header)
```

Comment:
curious if there is a reason to `put` instead of [`put_nowait`](https://docs.python.org/3/library/asyncio-queue.html#asyncio.Queue.put_nowait)

Reply:
The `await` seems cleaner than trying to catch a QueueFull and then wait for the queue to clear, then continue adding.